### DISCUSSION TITLE
Should we be supporting dynamic DAG generation in Prod?

### DISCUSSION THREAD

**Discussion by pnadolny13** [meltano/meltano#6559] [General]
Should we be supporting dynamic DAG generator in production?

The way that the airflow DAG generator works right now requires that we are constantly running cli commands and asking meltano for a list of schedules/jobs which has pros and cons.

- Pro: allow users to edit crons or other attributes of schedules in production via the UI, I dont know if we support this right now or if we want to but I can see why that would be useful. In that case a change to the meltano.yml might get updated by the UI so it would need to be constantly checked to pull in any schedule changes. Airflow's default is every 30 seconds. 
- Con: supporting dynamic DAGs means we run a meltano subprocess command every 30 seconds. This is a semi anti pattern from my interpretation of Airflow best practices since they recommend doing very little in the actual DAGs, although a quick subprocess isnt that much.
- Con: In our Squared production deployment on kubernetes we chose to use the default Airflow helm chart with the KubernetesPodOperator so the Airflow containers dont even have meltano installed to be able to run this command. I got around this by creating a static cache of the `schedule list` output https://github.com/meltano/squared/blob/b1bef50cc1d7c7e8bf0ffc79246f498e6b7bca49/data/orchestrate/dags/meltano.py#L210 and reading that in instead of calling meltano when in prod. This means that if I change a schedule using the meltano UI, its not persisted or pulled into Airflow.
- Con: dynamic edits to the meltano.yml in production dont get persisted to git or to images if youre using containers in prod. So if I redeploy or if my container restarts then these changes get wiped out. 

My first thought is that it goes against our vision of sending everything through git and code reviews if someone could click around a UI and make changes. Although a schedule is on the border line of something that might be acceptable to manipulate in prod if we had a way to get it persisted.

This is focused on the Airflow use case specifically but I think it would inform a larger pattern for deployment in general and whether we want to encourage dynamic production pipelines or not. We could come up with a caching pattern similar to what I had to do in the Squared project and recommend that as the default.

@aaronsteers @pandemicsyn @edgarrmondragon @tayloramurphy @kgpayne @alexmarple what do you guys think?

**Comment by tayloramurphy** [comment]
@pnadolny13 thanks for bringing this up. 

I'm inclined to say no we shouldn't allow changes to DAGs and their schedules in production. In this case, I think the read-only mode of the UI makes sense when the production deployment is activated. If users want to make a change, then they could enter an editing mode and either have that in a different environment or the same environment with a different git branch. 

It seems like [this part](https://github.com/meltano/files-airflow/blob/main/bundle/orchestrate/dags/meltano.py#L170-L177) of the meltano dag could be part of the deploy process for when prod is updated - which is essentially what you're advocating for with the cache mechanism. 

This makes me think about dbt too in that we could have compile step when something is deployed to production. The difference there is that there are dynamic incremental parts of a users dag within dbt models (incremental comes to mind) and we couldn't fully account for that. But the jobs and schedules as defined by Meltano should be deterministic on deploy.

**Reply by pnadolny13** [reply]
@tayloramurphy I agree with that workflow. What do you think about schedules? Like if I wanted to adjust my DAG to run at 8am instead of 9am, the best practice would still be to send it through a PR? I had this exact discussion in the past with my previous team and we ended up forcing everything through a PR but I do see the other point of view that argues a schedule isnt code. Although I can imagine a case where adjusting a schedule _could_ have downstream impacts for loosely coupled workflows like our push to S3 for the hub, if someone edited that schedule the file might not get refreshed before the daily hub build so it would be 1 day our of sync (that type of coupling is probably a bad practice though too).

I'm wondering if we should go ahead and implement this in the default dag generator so its more clear to users that our vision disallows editing anything in prod without a PR. We'd still ideally poll for changes when running locally otherwise we'll have to recommend users clear the cache file any time a schedule changes (not ideal). Also we'll need to document that the cache file should included in the build/deploy process. Maybe its a setting that defaults to poll but we recommend in the prod instructions that it be set to read from the cache.

**Reply by kgpayne** [reply]
> I had this exact discussion in the past with my previous team and we ended up forcing everything through a PR

Ditto 🙂 I think this is a reasonable approach right now, given a more technical (data engineering) target persona. Similar to dbt and Looker, we might choose to support authoring PR's via the UI experience in future, to allow things like schedule changes to be both UI-convenient and still in-line with PR workflow best practices. So the flow might be UI creates PR>merge triggers deploy>deploy changes schedule in UI.

One thing though about pushing _everything_ through a PR though is the long(ish) release cycle. We might choose to support fast-tracking minor changes like schedule updates to save users waiting for a full test suite run 🤔 We had a similar complaint with dbt at my previous place - minor changes (docs, typo fixes etc.) would have to wait 45mins+ to make it to prod because that's how long our test suits ran. We ended up giving some controls to analysts to fast-merge changes without review or tests.

**Reply by tayloramurphy** [reply]
@pnadolny13 still through a PR. I think we can be opinionated about that and give users a way to solve for some of those issues, but in my view, changing a schedule is a material change to the state of the system and should be reviewed.
